Add spatial hashing

fluidnumerics-joe opened the pull request:
<!--  The PR title should summarize the changes, for example "Add `Grid._build_face_dimension` function".
      Avoid non-descriptive titles such as "Addresses issue #229". -->

<!--  Replace XXX with the number of the issue that this PR will resolve. If this PR closed more than one,
      you may add a comma separated sequence-->
Closes #1126 

## Overview
<!--  Please provide a few bullet points summarizing the changes in this PR. This should include
      points on any bug fixes, new functions, or other changes that have been made. -->
This PR adds the `SpatialHash` class in the `neighbors.py` module. This class is used to encapsulate the necessary information and methods for conducting unstructured grid searches using spatial hashes. As part of the initialization of the `SpatialHash` object, an associated unstructured grid is used to construct a uniformly spaced structured grid, called the "hash grid", and relate cell indices in the hash grid to elements in the unstructured grid. This hash table is a critical element of the spatial hash grid search as it is used to create a short-list of elements to search within.

The `SpatialHash.query` method returns the element id that a point (or element ids that a list of points)  reside within alonside the barycentric coordinates. I've included helper methods to compute the barycentric coordinates of a point in relation to a given unstructured grid face (it works for convex faces n>=3 vertices). The barycentric coordinates are used to determine if a point is inside or outside a face; if all the barycentric coordinates are between [0,1], then a point  is inside a face.

Currently, only spherical coordinates (in radians or degrees) are supported for this search.

## Expected Usage

```Python
import uxarray as ux
grid_path = "../../test/meshfiles/fesom/soufflet-netcdf/grid.nc"
uxgrid = ux.open_grid(grid_path)

uxds = ux.open_dataset(grid_path, data_path)

# Construct the `SpatialHash` object
spatial_hash = uxgrid.get_spatial_hash(
    reconstruct="False",
)

# Query the `SpatialHash` for the face id and the barycentric coordinates.
face_ids, bcoords = spatial_hash.query([0.9, 1.8])

```

## PR Checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. If an entire section doesn't
apply to this PR, comment it out or delete it. -->

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [x] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [x] Adequate tests are created if there is new functionality
- [x] Tests cover all possible logical paths in your function
- [x] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [x] Docstrings have been added to all new functions
- [x] Docstrings have updated with any function changes
- [ ] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [ ] User functions have been added to `docs/user_api/index.rst`

**Examples**
- [x] Any new notebook examples added to `docs/examples/` folder
- [x] Clear the output of all cells before committing
- [x] New notebook files added to `docs/examples.rst` toctree
- [ ] New notebook files added to new entry in `docs/gallery.yml` with appropriate thumbnail photo in `docs/_static/thumbnails/`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

**PR Etiquette Reminders**
- This PR should be listed as a draft PR until you are ready to request reviewers

- After making changes in accordance with the reviews, re-request your reviewers

- Do *not* mark conversations as resolved if you didn't start them

- Do mark conversations as resolved *if you opened them* and are satisfied with the changes/discussion.
-->

philipc2 commented:
Hi @fluidnumerics-joe 

This looks great so far! May you add a `test_spatial_hasing.py` module with a few tests and run `pre-commit`? Let me know if you have any questions.

philipc2 reviewed (commented):
Can you add the `spatial-hashing` notebook here?  (both in the outline and the toc tree at the bottom)

I'm thinking between "Subsetting" and "Cross-Sections?" would be a good spot. 

https://github.com/UXARRAY/uxarray/blob/main/docs/userguide.rst?plain=1

fluidnumerics-joe commented:
> Hi @fluidnumerics-joe
> 
> This looks great so far! May you add a `test_spatial_hasing.py` module with a few tests and run `pre-commit`? Let me know if you have any questions.

Yes, I'll take care of both of these.

fluidnumerics-joe commented:
@philipc2 - I've added tests for the spatial hashing, fixed formatting, and added spatial hashing to the toc in the userguide. Let me know if there's anything else you'd like to see updated

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,15 @@ def get_kd_tree(
 
         return self._kd_tree
 
+    def get_spatialhash(
+        self,
+        reconstruct: bool = False,
+    ):
```
Please include a docstring, similar to how `get_kd_tree()` and `get_ball_three()` are formatted.

philipc2 replied:
Looks great!

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,207 @@ def coordinates(self, value):
+    reconstruct : bool, default=False
+        If true, reconstructs the tree
+
+    Notes
+    -----
+
```
Can remove if we don't plan to add anything to here.

fluidnumerics-joe replied:
Resolved in https://github.com/UXARRAY/uxarray/pull/1169/commits/d599276601dc2fedca6b9069070bc52702d579ac

philipc2 reviewed (changes requested):
Looks great! A few small comments attached.

I'm going to play around with this today.

Thanks for putting this together so quickly!

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,29 @@ def get_kd_tree(
 
         return self._kd_tree
 
+    def get_spatialhash(
```
```suggestion
    def get_spatial_hash(
```

I think this name is a better fit. It should fix the notebook too, since it's using this syntax.

![image](https://github.com/user-attachments/assets/4a459a24-257e-4d1b-a522-37d0e4162673)

philipc2 commented:
I'm having some issues with the `soufflet-netcdf/grid.nc` grid (seems to be crashing the docs too).

It stalls for multiple minutes when trying to construct the tree. This was one of the grids that gave me some trouble in #1013

philipc2 commented:
I get the following warning with it as well. Might be something wrong with how we're reading the file internally. 

> Latitude coordinates are not regularly spaced.
> Longitude coordinates are not regularly spaced.

fluidnumerics-joe commented:
> I'm having some issues with the `soufflet-netcdf/grid.nc` grid (seems to be crashing the docs too).
> 
> It stalls for multiple minutes when trying to construct the tree. This was one of the grids that gave me some trouble in #1013

I suspect this is related to the periodicity on a longitudinal extent of ~4.5 degrees ; it's an odd channel model that we've discussed with the Parcels and FESOM teams that I don't think we plan on supporting initially. Let me see if I can just get a rectangular ocean basin with closed boundaries.

fluidnumerics-joe commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,199 @@ def coordinates(self, value):
+            )
+            i2, j2 = self._hash_index2d(coords)
+
+            for eid in range(self._source_grid.n_face):
+                for j in range(j1[eid], j2[eid] + 1):
+                    for i in range(i1[eid], i2[eid] + 1):
+                        index_to_face[i + self._nx * j].append(eid)
+
+            return index_to_face
```
I'll get this added to the tests. Please provide a complete list of mesh files you will want to see tested. I'd like to try to be efficient about this.

philipc2 replied:
Sounds good!

That specific grid is too large to use in our tests. 

I think the files you have so far in the tests are sufficient. Maybe try the following MPAS one instead of the one you are currently using? 

https://github.com/UXARRAY/uxarray/blob/main/test/meshfiles/mpas/QU/oQU480.231010.nc

fluidnumerics-joe replied:
I'll add that and put in tests for both the dual and the primal grid.

fluidnumerics-joe replied:
I've got the mpas dual and primal grid tests added in https://github.com/UXARRAY/uxarray/pull/1169/commits/0c0dd6021c04569d7a055450beba63c87d768000 . This was a good test; I ended up switching to the [generalized Waschpress points](https://www.mn.uio.no/math/english/people/aca/michaelf/papers/wach_mv.pdf) for the barycentric coordinates so that they could be used for barycentric interpolation. 

For that larger grid that failed, I'd say we open that as a separate issue to be resolved after this PR. I suspect the resolution could be handled by padding out the hash grid, though until I look at that mesh, I won't be certain.

fluidnumerics-joe replied:
I've switched to the mpas example in the notebook in https://github.com/UXARRAY/uxarray/pull/1169/commits/60e670ae79d9d4a1c921516f7023b70c412ab4ee

I think this resolves this request. Feel free to re-open if not.

philipc2 replied:
Looks to be good now! Thanks!

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,29 @@ def get_kd_tree(
+        Parameters
+        ----------
+        reconstruct : bool, default=False
+            If true, reconstructs the tree
```
```suggestion
            If true, reconstructs the spatial hash
```

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,29 @@ def get_kd_tree(
+        -------
+        self._spatialhash : grid.Neighbors.SpatialHash
+            SpatialHash instance
+        """
```
It would be nice to have a minimal example in the docstring here, showing how to query the data structure. Something like this adapted from the notebook could be good. 

```
Examples
--------
>>> import uxarray as ux
Open a grid from a file path
>>> uxgrid = ux.open_grid("grid_filename.nc")

Obtain SpatialHash instance
>>> spatial_hash = uxgrid.get_spatial_hash()

Query to find the face a point lies within in addition to its barycentric coordinates.
>>> face_ids, bcoords = spatial_hash.query([0.0, 0.0])
```

fluidnumerics-joe replied:
Added this to the docstring in https://github.com/UXARRAY/uxarray/pull/1169/commits/0c0dd6021c04569d7a055450beba63c87d768000

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,41 @@ def get_kd_tree(
 
         return self._kd_tree
 
+    def get_spatial_hash(
```
Please include this in [`api.rst`](https://github.com/UXARRAY/uxarray/blob/main/docs/api.rst)

fluidnumerics-joe replied:
I couldn't find an appropriate spot; `get_kd_tree` and `get_ball_tree` are not listed in there anywhere and these would be the most similar "search" methods..

philipc2 replied:
Oh that's a good catch; we need to include those.

Feel free to skip it for now, I'll open up a PR and include all three with appropriate sections after this is merged.

fluidnumerics-joe replied:
Added to Grid/Methods section in api.rst in https://github.com/UXARRAY/uxarray/pull/1169/commits/b655a46cc9e699bbee43fd1b9561539d72bb2edd

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1749,6 +1751,41 @@ def get_kd_tree(
+
+        Examples
+        --------
+        >>> import uxarray as ux
+        Open a grid from a file path
+        >>> uxgrid = ux.open_grid("grid_filename.nc")
+
+        Obtain SpatialHash instance
+        >>> spatial_hash = uxgrid.get_spatial_hash()
+
+        Query to find the face a point lies within in addition to its barycentric coordinates.
+        >>> face_ids, bcoords = spatial_hash.query([0.0, 0.0])
+        """
```
```suggestion
        Open a grid from a file path:
         
        >>> import uxarray as ux
        >>> uxgrid = ux.open_grid("grid_filename.nc")

        Obtain SpatialHash instance:
        
        >>> spatial_hash = uxgrid.get_spatial_hash()

        Query to find the face a point lies within in addition to its barycentric coordinates:
        
        >>> face_ids, bcoords = spatial_hash.query([0.0, 0.0])
        """
```

philipc2 commented on docs/user-guide/spatial-hashing.ipynb:
```diff
@@ -0,0 +1,104 @@
+{
```
<p>"UXArray" -&gt; "UXarray"</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UXARRAY/uxarray/pull/1169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

fluidnumerics-joe replied:
Fixed in https://github.com/UXARRAY/uxarray/pull/1169/commits/fc59591fae435817c73a24ecaf83be0c91ce0998

philipc2 commented on docs/user-guide/spatial-hashing.ipynb:
```diff
@@ -0,0 +1,104 @@
+{
```
<p>Should be updated. To keep things simple, maybe we can use our quad hexagon grid, which is a small example grid. </p><p> </p>

<pre class="ql-syntax" spellcheck="false">grid_path = "../../test/meshfiles/ugrid/quad-hexagon/grid.nc"
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UXARRAY/uxarray/pull/1169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

philipc2 replied:
It's very tiny

<img width="774" alt="image" src="https://github.com/user-attachments/assets/1edfbc53-3b99-4809-8608-8324a6c1eb64" />


Feel free to keep the MPAS one.

fluidnumerics-joe replied:
Updated to smaller hex grid here : https://github.com/UXARRAY/uxarray/pull/1169/commits/fb4bbf82a0647c262a113b5a5a0dfa99385e0c23

philipc2 commented on docs/user-guide/spatial-hashing.ipynb:
```diff
@@ -0,0 +1,104 @@
+{
```
<p>"UXArray" -&gt; "UXarray" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UXARRAY/uxarray/pull/1169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

fluidnumerics-joe replied:
Fixed in https://github.com/UXARRAY/uxarray/pull/1169/commits/fc59591fae435817c73a24ecaf83be0c91ce0998

philipc2 commented on docs/user-guide/spatial-hashing.ipynb:
```diff
@@ -0,0 +1,104 @@
+{
```
<p>Can we print the face indices and/or coordinates here to showcase what the output looks like? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UXARRAY/uxarray/pull/1169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

fluidnumerics-joe replied:
I can do that, but this checkbox in the PR template conflicts with this request : " Clear the output of all cells before committing"

fluidnumerics-joe replied:
Output is retained in https://github.com/UXARRAY/uxarray/pull/1169/commits/fc59591fae435817c73a24ecaf83be0c91ce0998

philipc2 replied:
The wording is pretty confusing, it's intended to mean that the notebook should be restarted and cleared before committing. 

I think our default template is due for an overhaul anyways, thanks for noting this.

fluidnumerics-joe commented:
> Looks like one of the Github suggestions broke the pre-commit. Can you run it again?
> 
> Also, please restart and clear the outputs of the notebook, they will be rendered via our CI. Other than that all looks good.

I re-ran `pre-commit run --all-files` and cleared python notebook output.

aaronzedwick commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+    """
+    Compute the signed area of a triangle given by three points.
+    """
+    return 0.5 * abs(A[0] * (B[1] - C[1]) + B[0] * (C[1] - A[1]) + C[0] * (A[1] - B[1]))
```
Quick question about this: If it is computing the signed area of the triangle, would the `abs` call cause an issue?

fluidnumerics-joe replied:
Fair enough. This docstring is not accurate. I've moved away from computing signed area in favor of an error check on the coordinates using barycentric interpolation. I'll correct the docstring

fluidnumerics-joe replied:
Corrected in https://github.com/UXARRAY/uxarray/pull/1169/commits/e7ab0316d9b30b61ddfd01a7998e88e9d5359742

philipc2 commented:
> > Looks like one of the Github suggestions broke the pre-commit. Can you run it again?
> > Also, please restart and clear the outputs of the notebook, they will be rendered via our CI. Other than that all looks good.
> 
> I re-ran `pre-commit run --all-files` and cleared python notebook output.

Oops, look's like its complaining about a merge conflict.

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+
+        return faces, bcoords
+
+
```
```suggestion
@njit(cache=True)
```

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+    """
+    return 0.5 * abs(A[0] * (B[1] - C[1]) + B[0] * (C[1] - A[1]) + C[0] * (A[1] - B[1]))
+
+
```
```suggestion
@njit(cache=True)
```

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+        a2 = _triangle_area(point, vi, vi1)
+        sum_wi += a0 / (a1 * a2)
+        w.append(a0 / (a1 * a2))
+
```
For some queries, I'm received a divide by zero error. Can we add a small value to the sum before dividing?

We can import the `ERROR_TOLLERANCE` from `uxarray.constants`

fluidnumerics-joe replied:
Can you share a specific example that is causing this problem so that we can get it added to the tests ? I'd like to make sure whatever patch we put it in here is covered in the test cases .

philipc2 replied:
The error arose when I decorated the function with Numba. In it's current state, it raises warnings when using the 30km MPAS grid I tested above, but I haven't run into it in other instances. 

We could extend the existing tests and make sure they don't raise any warnings in the parts this function is called.

fluidnumerics-joe replied:
I've put in the following in the barycentric weights calculation in https://github.com/FluidNumerics/uxarray/commit/378c817ff5911abda8f28679c77671c6031a9ff5

```
        sum_wi += a0 / (max(a1 * a2, ERROR_TOLERANCE))
        w.append(a0 / (max(a1 * a2, ERROR_TOLERANCE)))
```

This clips the denominator to be no smaller than `ERROR_TOLERANCE` from `uxarray.constants` . 

I did not put in an additional test. I'm not familiar with which exact mesh file and which query points you used to produce the error. For a triangle area to be zero (either `a1` or `a2` to be zero), I suspect that a point must lie on an edge of a face or exactly on a corner node. I'm cooking up an example for that at the moment.

philipc2 replied:
This is perfect. 

Once you merge the example, feel free to merge, as everything else looks good. We're excited to have this part of our release today.

fluidnumerics-joe replied:
https://github.com/UXARRAY/uxarray/pull/1169/commits/b3c9a0761217f8efc34e57bfd77ffd2d16f03b86 adds tests for querying on vertices and edges in a simple scenario. In this, I found it was needed to clip the individual triangle areas, rather than the denominator as a whole, to get "query on a vertex" to compute the correct value. This should resolve division by zero errors you were encountering with MPAS 30km grid. Once you confirm that, please resolve this conversation so we can get this merged in.

philipc2 commented on uxarray/grid/neighbors.py:
```diff
@@ -781,6 +781,204 @@ def coordinates(self, value):
+
+            return index_to_face
+
+    def query(
```
For a 30km MPAS grid, it was taking about 6-7 seconds for each query. Here's a slightly revised version that I was able to get down to about 0.5s per query. 

I'm happy to help explore further optimizations, since I assume you'll want to perform a large number of queries.

```Python
coords = _prepare_xy_for_query(coords, in_radians, distance_metric=None)
num_coords = coords.shape[0]
max_nodes = self._source_grid.n_max_face_nodes

# Preallocate results
bcoords = np.zeros((num_coords, max_nodes), dtype=np.double)
faces = np.full(num_coords, -1, dtype=INT_DTYPE)

# Get grid variables 
n_nodes_per_face = self._source_grid.n_nodes_per_face.to_numpy() 
face_node_connectivity = self._source_grid.face_node_connectivity.to_numpy()

# Precompute radian values for node coordinates:
node_lon = np.deg2rad(self._source_grid.node_lon.to_numpy())
node_lat = np.deg2rad(self._source_grid.node_lat.to_numpy())

# Get the list of candidate faces for each coordinate
candidate_faces = [self._face_hash_table[pid] for pid in self._hash_index(coords)]

for i, (coord, candidates) in enumerate(zip(coords, candidate_faces)):
    for face_id in candidates:
        n_nodes = n_nodes_per_face[face_id]
        node_ids = face_node_connectivity[face_id, :n_nodes]
        nodes = np.column_stack((node_lon[node_ids], node_lat[node_ids]))
        bcoord = np.asarray(_barycentric_coordinates(nodes, coord))
        err = abs(np.dot(bcoord, nodes[:, 0]) - coord[0]) + abs(np.dot(bcoord, nodes[:, 1]) - coord[1])
        if (bcoord >= 0).all() and err < tol:
            faces[i] = face_id
            bcoords[i, :n_nodes] = bcoord[:n_nodes]
            break  

return faces, bcoords
```

fluidnumerics-joe replied:
I'm happy to implement this change you've shared here, but I think we should address additional performance optimizations in a future PR. Let's not keep moving the goal post here.

fluidnumerics-joe replied:
In practice, for a lagrangian code like Parcels, the spatial_hash will be used to initially populate the face ids for particle locations. Since we're mostly using explicit time integration methods, lookups for face ids in future time steps will be done using connectivity tables in uxarray (since a particle likely won't have a CFL > 1 ). When particles are not found by means of looking through the current associated face id, or the neighbors, the query will be run as a last ditch effort before determining that a particle is lost.

In our view, functionality is most critical at this point. Once we have concrete examples with benchmark data and user feedback, it's possible performance may need to be addressed. But right now, it's a "down-the-road" concern.

philipc2 replied:
> I'm happy to implement this change you've shared here, but I think we should address additional performance optimizations in a future PR. Let's not keep moving the goal post here.

This works with me. I'm happy to continue iterating on performance once this is merged. If you are happy with my suggestion and would like to implement it, go ahead!

> In our view, functionality is most critical at this point. Once we have concrete examples with benchmark data and user feedback, it's possible performance may need to be addressed. But right now, it's a "down-the-road" concern.

Sounds good! Other than the performance suggestions, I am very pleased with the implementation. Great work!

fluidnumerics-joe replied:
You're changes were added in https://github.com/UXARRAY/uxarray/pull/1169/commits/5fcd039512f209149bf7d5cf9c921874ca1f9d82

philipc2 replied:
Thanks!

philipc2 reviewed (approved):
Thanks for this contribution! Other than the comments I left earlier, it looks good to me.

aaronzedwick reviewed (approved):
This looks great! I will actually make use of the barycentric weights in one of my current PR's, so that is really helpful to me. Thanks for your contribution!

philipc2 commented:
@fluidnumerics-joe 

Turns out the issue with the `30km` MPAS grid was the `edge_node_distances` were represented in `meters`, instead of `radians`, which is inconsistent with our `Grid.edge_node_distances` definition. The same behavior can be see in the `meshfiles\mpas\QU\oQU480.231010.nc` grid, but the result is still correct. This was causing the hash cell size to be much larger than exepcted. 

It also appears we have a typo in our `Grid.edge_node_distances` definition. The unit should be `radians` instead of `degrees` here 

https://github.com/UXARRAY/uxarray/blob/d9f091f6223e5881b3b3a5784ba4bd0869b20437/uxarray/grid/grid.py#L1386-L1394

I'll raise an issue and get this fixed. Glad to see that the issue wasn't anything with your implementation.

fluidnumerics-joe commented:
🥳 - thanks @philipc2 and @aaronzedwick for your feedback. Without it, I would have not seen some of the cases that we ended up building tests for. This is great to see pushed across the finish line. 

Feel free to reach out on any future issues related to the spatial hashing. I'm happy to chip in.

rajeeja commented:
@fluidnumerics-joe 

Thanks for this work and hoping to see few more features in future :)
